### DISCUSSION TITLE
Nvidia issues and workaround

### DISCUSSION THREAD

**Discussion by CicadaSeventeen** [SpiralLinux/SpiralLinux-project#146] [General]
Due to some bugs, now installing nVidia driver on Spiral need some tricks. The bug should be fixed in the next version. Here is the workaround.
```
sudo apt update
sudo apt upgrade
sudo apt remove "nvidia*"
sudo apt install -t bullseye --install-recommends nvidia-driver
```
Please do not install `nvidia-xconfig`

**Comment by Senshi111** [comment]
Just to tell that i manually installed latest nvidia 515.76 on spiral and it works great for me.

**Reply by CicadaSeventeen** [reply]
So how can you do that?

**Reply by CicadaSeventeen** [reply]
You use driver from nvidia rather than debian repository?

**Reply by Senshi111** [reply]
Standard download latest nvidia from official site install some dependencies that's needed for 32-bit libs, and that's all 
```
linux-headers-$(uname -r) build-essential libglvnd-dev pkg-config
```

**Reply by CicadaSeventeen** [reply]
BTW, the lastest driver may break RTX 3060, knowing from arch users

**Reply by Senshi111** [reply]
I have 3070ti

**Reply by CicadaSeventeen** [reply]
> Just to tell that i manually installed latest nvidia 515.76 on spiral and it works great for me.

Could you tell me what kernel you use now?

**Reply by Senshi111** [reply]
> > Just to tell that i manually installed latest nvidia 515.76 on spiral and it works great for me.
> 
> Could you tell me what kernel you use now?

Sure. Xanmod 6.0

**Reply by CicadaSeventeen** [reply]
> > > Just to tell that i manually installed latest nvidia 515.76 on spiral and it works great for me.
> > 
> > 
> > Could you tell me what kernel you use now?
> 
> Sure. Xanmod 6.0

I find that that Xanmod repository actually use different gcc version from debian stable. It can cause problem, especially with nvidia driver, take care

**Reply by Senshi111** [reply]
> > > > Just to tell that i manually installed latest nvidia 515.76 on spiral and it works great for me.
> > > 
> > > 
> > > Could you tell me what kernel you use now?
> > 
> > Sure. Xanmod 6.0
> 
> I find that that Xanmod repository actually use different gcc version from debian stable. It can cause problem, especially with nvidia driver, take care

Im not on debian stable, but on sid

**Comment by CicadaSeventeen** [comment]
~~Now it seems that nvidia driver only work for default kernel but not 5.10 nor custom kernel~~
Solved.
Scripts here may not work because I usually make awful bugs. 
[Scripts.tar.gz](https://github.com/SpiralLinux/SpiralLinux-project/files/9762291/Scripts.tar.gz)

If so, please build dkms your self and then update initrafs.

**Reply by geckolinux** [reply]
> Now it seems that nvidia driver only work for default kernel but not 5.10 nor custom kernel

I'm sorry I didn't understand this, the default Debian Stable kernel is 5.10. Did you mean it doesn't work with kernel 5.18?

**Reply by CicadaSeventeen** [reply]
While Debian use 5.10 as default kernel, an out-of-box SpiralLinux use 5.18 as the default kernel when booting. If I use debian nonfree repository package `nvidia-dirvier` to install nv driver, then only 5.18 kernel works, while other kernel like 5.10 cannot load Nvidia driver.
I guess because debian does not have a good multi kernel support as opensuse,  The official post-install script only generate booting config for "the first kernel". I try to use `update-initramfs -k all -u` but not work. I guess there is something else needed to be regenerate.

**Reply by geckolinux** [reply]
> I guess because debian does not have a good multi kernel support as opensuse

Correct. I would recommend to pick which kernel you want to use, uninstall the other one, boot into your chosen kernel, and then try to install the Nvidia driver.

**Reply by CicadaSeventeen** [reply]
> > I guess because debian does not have a good multi kernel support as opensuse
> 
> Correct. I would recommend to pick which kernel you want to use, uninstall the other one, boot into your chosen kernel, and then try to install the Nvidia driver.

I have managed to fix this problem. And I write `.sh` script for those who want to use multi kernel with Nvidia. But I do not try script myself, and I am bad at coding, so I cannot say whether my script works.
But the point is like:
```
dkms build -k 5.10.0-18-amd64 -m nvidia-current -v 470.141.03
dkms autoinstall
update-initramfs -k all -u
```
 Kernel Name can Nvidia driver version may change in future.

For those using custom kernel, maybe need ENV as "IGNORE_CC_MISMATCH=1". This means someone uses wrong version to build kernel. So not promised to build dkms in such condition, but it works for me. (I want 5.15 but too lazy to build it myself so I steal some strange Xanmod 5.15 kernel built by gcc11)
[Scripts.tar.gz](https://github.com/SpiralLinux/SpiralLinux-project/files/9762283/Scripts.tar.gz)

**Comment by CicadaSeventeen** [comment]
Known issue: may get black screen after suspend and resume.
Hibernate is OK.

**Comment by geckolinux** [comment]
Thanks for the tests. So, to summarize, a few questions:

1. Is there any way to install the Nvidia driver without special scripts on a default SpiralLinux installation?
2. Is it possible to install the Nvidia driver without special scripts on a default SpiralLinux installation **_with only the 5.10 kernel installed_**?
3. Is it possible to install the Nvidia driver without special scripts on a default SpiralLinux installation **_with only the 5.18 kernel installed_**?

**Reply by CicadaSeventeen** [reply]
Answer of question 1: yes but only for kernel 5.18, if you switch to kernel 5.10 when booting, nvidia driver not works then.

Answer of question 2and 3: I do not really know but very likely to be yes. I may help to check sometimes but it will force me to reinstall a system so I will not test it so far.

**Reply by geckolinux** [reply]
1. What about if you remove the APT pinning for the Nvidia packages?

Re: 2 and 3:
>  I may help to check sometimes but it will force me to reinstall a system so I will not test it so far.

It would not require reinstalling the system. Just remove everything related to kernel 5.10 and then test Nvidia installation with 5.18, and then remove everything related to kernel 5.18 and install Nvidia packages with kernel 5.10. You could also take a Snapper snapshot before testing.

**Comment by brakenow** [comment]
Didn't work with the latest 6.0 kernel I updated to after a fresh install. Boot hanged with the two last lines being cryptsetup successfully opened the volumes.

**Reply by geckolinux** [reply]
Hi, what exactly didn't work? The boot was hanging indefinitely?

**Reply by Senshi111** [reply]
Are you using kde or?

**Reply by CicadaSeventeen** [reply]
It is normal that Nvidia driver from Debian Bullseye does not work for kernel 6.x. Actually a certain version of nvidia driver only supports some version of kernels and kernel 6.x needs more advanced nvidia driver version, which is not included from Bullseye repos. I do not know whether this is one in backports, but we have found out that Nvidia drivers form Bullseye-backports cannot installed out of box.
I have known such problems from Arch and Sid users, but I have not told you because I get depressed these days once again. Sorry.
I suggest those who does not need most advanced kernel to stay at 5.19, or htey can just install Nvidia driver from Nvidia official site rather than from Debian's repository.

**Reply by brakenow** [reply]
> Hi, what exactly didn't work? The boot was hanging indefinitely?

Hi, yes.

**Reply by brakenow** [reply]
> Are you using kde or?

Gnome

**Reply by Senshi111** [reply]
Maybe the best way for you is to install nvidia manually

**Reply by geckolinux** [reply]
> we have found out that Nvidia drivers form Bullseye-backports cannot installed out of box.

I wonder if this was possibly fixed in Bullseye-backports since we last tested.

@brakenow  Can you please confirm how you installed the Nvidia drivers?

**Reply by brakenow** [reply]
> @brakenow Can you please confirm how you installed the Nvidia drivers?

First post in this thread.

**Reply by geckolinux** [reply]
OK, @CicadaSeventeen knows a lot more about Nvidia on Debian than I do, but wouldn't expect those instructions to work with the Backports kernel version. It would be good to confirm if the current Nvidia packages in Backports have been fixed to work with the latest Backports kernel (6.0).

**Reply by Senshi111** [reply]
As i remember nvidia versions <520. Needs to be patch to be able to work with kernel 6+
That's what i had to do when i was using kernel 6rc with at thet time latest nvidia i think it was 515.76

**Comment by marksalendu** [comment]
Hi, I'm noob here. So apology, if my question is too dumb.

If I have laptop comes with nvidia gpu, will it work if I just install spiral linux without installing any additional driver?
I use this laptop not for gaming, only for browsing, watch youtube, some office works.

**Reply by Senshi111** [reply]
I don't think you would have any problems. It should work out of the box.
I had it on my laptop but i did install nvidia and some firmware for my 12th gen Intel cpu. Because i used it for gaming also

**Reply by geckolinux** [reply]
Hi there @marksalendu , yes it will probably work fine out of the box. I also have a laptop with integrated Intel graphics plus a more power Nvidia card, but I don't need or want the Nvidia part of it and SpiralLinux works fine with just the Intel graphics.

**Reply by marksalendu** [reply]
Thanks for the reply. Will test it out.

**Comment by nomadnewbie** [comment]
Hello everyone, I just wanted to post my 2 cents here as I was having trouble with installing NVIDIA also for, well since I first installed SpiralLinux. However, that has been solved just recently when I upgraded my Kernel to 6. I did nothing special, just removed all previous and any left over NVIDIA code and purged before rebooting, praying I was not going to get a "black screen"l and I did not, and I immediately preceded to install NVIDIA, and for the first time in months it installed without a hitch this time. Here is my system and specs: Spiral Flavor
Debian GNU/Linux 11 (bullseye) x86_64
Kernel: 6.0.0-0.deb11.2-amd64
ROG Strix GA35DX_G35DX 1.0
GNOME 3.38.6
AMD Ryzen 7 5800X (16) @ 4.100GHz
NVIDIA GeForce RTX 3080
NVIDIA-Linux-x86_64-525.60.11.run
yes, I tried the latest NVIDIA driver at the time because all the others failed, even the previous one I had before installing Spiral and which I had no problems with when I tried out MX. So, I cannot say why this driver worked, if it was the Kernel, or perhaps NVIDIA fixed whatever was causing the other drivers to not install, but this one worked and still is and I have had absolutely no problems with it since I installed it. So, good luck to those still having problems if this one does not work for you.

**Reply by geckolinux** [reply]
Great to hear that @nomadnewbie , thanks for reporting. So you didn't have to change any of the SpiralLinux default  APT pinning in `/apt/preferences.d/99backports` ?

**Reply by Senshi111** [reply]
Every time i was getting black screen when I'm installing nvidia is when i checked yes for question "do you want automatically to update you x configuration...". So i always say no to that, and everything is fine then

**Reply by geckolinux** [reply]
@Senshi111 Interesting, thanks. So the question is, for future versions do we keep the APT pin preference for `nvidia*` packages in Backports?

**Comment by nomadnewbie** [comment]
No, I did absolutely nothing, I was just praying it was not going to make me have to reinstall all over again.Sent from my iPhoneOn Jan 1, 2023, at 3:30 PM, GeckoLinux ***@***.***> wrote:﻿
Great to hear that @nomadnewbie , thanks for reporting. So you didn't have to change any of the SpiralLinux default  APT pinning in /apt/preferences.d/99backports ?
